### PR TITLE
[v6r20] FTS3: add a protection against race condition between RMS and FTS3

### PR DESCRIPTION
**Pull request by chaen**
BEGINRELEASENOTES

*DMS
FIX: add a protection against race condition between RMS and FTS3 

ENDRELEASENOTES

### CODE REVIEW

File: DataManagementSystem/Agent/RequestOperations/ReplicateAndRegister.py [hunk 1]
```diff
@@ -101,9 +109,17 @@ def filterReplicas(opFile, logger=None, dataManager=None):
 
   for repSEName in replicas:
     repSEMetadata = StorageElement(repSEName).getFileMetadata(opFile.LFN)
-    error = repSEMetadata.get('Message', repSEMetadata.get('Value', {}).get('Failed', {}).get(opFile.LFN))
```
Comment (andresailer) [inline DataManagementSystem/Agent/RequestOperations/ReplicateAndRegister.py]: This line is less than 120 characters, why is this modified?
```
>>> len("    error = repSEMetadata.get('Message', repSEMetadata.get('Value', {}).get('Failed', {}).get(opFile.LFN))")
106
```

File: DataManagementSystem/Agent/RequestOperations/ReplicateAndRegister.py [hunk 2]
```diff
@@ -390,11 +407,80 @@ def ftsTransfer(self):
     # Just in case some transfers could not be scheduled, try them with RM
     return self.dmTransfer(fromFTS=True)
 
+  def _checkExistingFTS3Operations(self):
+    """
+       TL;DR: check if there are ongoing FTS3Operation for the current RMS Operation
```
Comment (andresailer) [inline DataManagementSystem/Agent/RequestOperations/ReplicateAndRegister.py]: The first line of the docstring is by convention a TL;DR

File: DataManagementSystem/Agent/RequestOperations/ReplicateAndRegister.py (hunk 1 above)
Comment (chaen) [reply DataManagementSystem/Agent/RequestOperations/ReplicateAndRegister.py]: I have no idea ! But it does it consistently

File: DataManagementSystem/Agent/RequestOperations/ReplicateAndRegister.py (hunk 2 above)
Comment (chaen) [reply DataManagementSystem/Agent/RequestOperations/ReplicateAndRegister.py]: OK, changed

File: DataManagementSystem/Service/FTS3ManagerHandler.py [hunk 3]
```diff
@@ -31,105 +33,128 @@ class FTS3ManagerHandler( RequestHandler ):
...
+
+  types_getOperationsFromRMSOpID = [(long, int)]
+
+  @classmethod
+  def export_getOperationsFromRMSOpID(cls, rmsOpID):
+    """ Get the FTS3Operation associated to a given rmsOpID
+
+        :param rmsOpID: ID of the operation in the RMS
+
+        :return: list of the FTS3Operations JSON matching
```
Comment (andresailer) [inline DataManagementSystem/Service/FTS3ManagerHandler.py]: Does this mean "JSON encoded list of FTS3Operations", the trailing "matching" is a bit confusing.

File: DataManagementSystem/Agent/RequestOperations/ReplicateAndRegister.py [hunk 4]
```diff
@@ -390,11 +395,80 @@ def ftsTransfer(self):
...
+    if not nonFinishedFTSOperations:
+      self.log.debug("No ongoing FTSOperations, all good")
+      return S_OK(True)
+
+    self.log.warn("Some FTS3Operations already exist for the RMS Operation:",
+                  [op.operationID for op in nonFinishedFTSOperations])
+
+    # This would really be a screwed up situation !
+    if len(nonFinishedFTSOperations) > 1:
+      self.log.warn("That's a serious problem !!")
```
Comment (fstagni) [inline DataManagementSystem/Agent/RequestOperations/ReplicateAndRegister.py]: then why only a warning?

File: DataManagementSystem/Agent/RequestOperations/ReplicateAndRegister.py [hunk 5]
```diff
@@ -390,11 +395,80 @@ def ftsTransfer(self):
...
 
     self.log.info("scheduling files in FTS3...")
 
+    # Check first if we do not have ongoing transfers
+
+    res = self._checkExistingFTS3Operations()
+    if not res['OK']:
+      return res
+
+    # if res['Value'] is False
```
Comment (fstagni) [inline DataManagementSystem/Agent/RequestOperations/ReplicateAndRegister.py]: ...this is not exactly the test you have below:
`if res['Value'] is False` is different from `if not res['Value']:`

**Comment by fstagni** [comment]
Don't you have any test to add or modify together with this PR?

File: DataManagementSystem/Agent/RequestOperations/ReplicateAndRegister.py (hunk 2 above)
Comment (chaen) [reply DataManagementSystem/Agent/RequestOperations/ReplicateAndRegister.py]: Oops

File: DataManagementSystem/Agent/RequestOperations/ReplicateAndRegister.py (hunk 4 above)
Comment (chaen) [reply DataManagementSystem/Agent/RequestOperations/ReplicateAndRegister.py]: because the error is not really here, I don't even know how it could happen. I could remove this altogether

File: DataManagementSystem/Agent/RequestOperations/ReplicateAndRegister.py (hunk 5 above)
Comment (chaen) [reply DataManagementSystem/Agent/RequestOperations/ReplicateAndRegister.py]: Correct, but it does not really matter here. But I can change the test, it does not harm

File: DataManagementSystem/Agent/RequestOperations/ReplicateAndRegister.py [hunk 6]
```diff
@@ -390,11 +407,80 @@ def ftsTransfer(self):
...
+    if not nonFinishedFTSOperations:
+      self.log.debug("No ongoing FTSOperations, all good")
+      return S_OK(True)
+
+    self.log.warn("Some FTS3Operations already exist for the RMS Operation:",
+                  [op.operationID for op in nonFinishedFTSOperations])
+
+    # This would really be a screwed up situation !
+    if len(nonFinishedFTSOperations) > 1:
+      self.log.warn("That's a serious problem !!")
```
[the remark this answers is not in the document]
Comment (chaen) [reply DataManagementSystem/Agent/RequestOperations/ReplicateAndRegister.py]: I think I just remove this altogether, since I don't even know how it could happen. It's more an `assert` that should be here...

File: DataManagementSystem/Agent/RequestOperations/ReplicateAndRegister.py [hunk 7]
```diff
@@ -390,11 +395,80 @@ def ftsTransfer(self):
...
+    if not res['OK']:
+      self.log.debug(
+          "Could not get FTS3Operations matching OperationID",
+          self.operation.OperationID)
+      return res
+
+    existingFTSOperations = res['Value']
+    # It is ok to have FTS Operations in a final state, so we
+    # care only about the others
+    nonFinishedFTSOperations = [
```
[the remark this answers is not in the document]
Comment (chaen) [reply DataManagementSystem/Agent/RequestOperations/ReplicateAndRegister.py]: thanks

File: DataManagementSystem/Agent/RequestOperations/ReplicateAndRegister.py [hunk 8]
```diff
@@ -390,11 +395,80 @@ def ftsTransfer(self):
...
+    # We take the rmsFileID of the files in the Operations,
+    # find the corresponding File object, and set them scheduled
+    rmsFileIDsToSetSchedule = set(
+        [ftsFile.rmsFileID for ftsOp in nonFinishedFTSOperations for ftsFile in ftsOp.ftsFiles])
+
+    for opFile in self.operation:
+      # If it is in the DB, it has a FileID
+      opFileID = opFile.FileID
+      if opFileID in rmsFileIDsToSetSchedule:
+        self.log.warn("Setting FTSFile as already scheduled", opFileID)
```
[the remark this answers is not in the document]
Comment (chaen) [reply DataManagementSystem/Agent/RequestOperations/ReplicateAndRegister.py]: Obviously...

File: DataManagementSystem/Agent/RequestOperations/ReplicateAndRegister.py [hunk 9]
```diff
@@ -390,11 +395,80 @@ def ftsTransfer(self):
...
+    self.log.warn("Some FTS3Operations already exist for the RMS Operation:",
+                  [op.operationID for op in nonFinishedFTSOperations])
+
+    # This would really be a screwed up situation !
+    if len(nonFinishedFTSOperations) > 1:
+      self.log.warn("That's a serious problem !!")
+
+    # We take the rmsFileID of the files in the Operations,
+    # find the corresponding File object, and set them scheduled
+    rmsFileIDsToSetSchedule = set(
```
[the remark this answers is not in the document]
Comment (chaen) [reply DataManagementSystem/Agent/RequestOperations/ReplicateAndRegister.py]: thanks

File: DataManagementSystem/Service/FTS3ManagerHandler.py (hunk 3 above)
Comment (chaen) [reply DataManagementSystem/Service/FTS3ManagerHandler.py]: yes, I change

**Comment by chaen** [comment]
Currently there are no tests for the integration between RMS and FTS. This is clearly missing and I intend to do it, but not in this PR, because it will take me too much time, and I don't want to risk loosing the hotfix

**Comment by chaen** [comment]
I removed the WIP.

@andresailer @sfayer if this gets merged, please do not forget to create the index on the rmsOpID on the Operations table

**Comment by atsareg** [comment]
I am going to merge this PR. This will need a special note in the DIRAC Wiki to update the database schema for the rmsOpID index. Is it just optimization or it will not work at all without the index ?

**Comment by chaen** [comment]
Eventually the query is going to take long without the index, but it is not mandatory... I'll add a note in the wiki

**Comment by atsareg** [comment]
Please, check this one: https://github.com/DIRACGrid/DIRAC/wiki/DIRAC-v6r20p8

**Comment by chaen** [comment]
perfect ! Thanks
